src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -2207,6 +2207,7 @@ private void IntrospectiveSort(int left, int length)
                 }
             }
 
+            [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Can we include a comment as to why this is here?  I imagine the calculus around whether this is valuable could change in the future.

Reply:
Added. 

Note `IntroSort` is nominally doubly recursive but has already had the tail recursive call transformed into a loop manually. If we were ever tempted to undo that and rely on the JIT to do something similar, we might find the `NoInline` annotation would inhibit that transformation too.

@EgorBo we might consider special heuristics for recursion, generally if the method is large and doesn't have a frequent fast path that avoids recursion, then there's not much to be gained.